### ISSUE TITLE
Copying the same DLL from more than one source & check for correct arch

### ISSUE THREAD

**Issue by Athanasius** [py2exe/py2exe#64]
We had a headscratcher today, trying to figure out why the GitHub build of our application resulted in a broken install, when it worked perfectly fine from source, and even when built locally.

We finally tracked this down to how py2exe does both of these:

```
Copy DLL C:\hostedtoolcache\windows\Python\3.7.9\x86\DLLs\libcrypto-1_1.dll to dist.win32\
Copy DLL C:\Program Files\Amazon\AWSCLIV2\libcrypto-1_1.dll to dist.win32\
```

When we checked a prior build it had also done this double-source copy, but in the other order, so we ended up with the correct version of the file.  That 'Amazon' DLL is a 64-bit one so plain doesn't work with our 32-bit program.

So, two issues:

1. Why does py2exe copy a DLL twice, from a different source each time ?  Surely it should either pick one and stick to it, or alert you that you need to make a decision.  Especially important as in some sense the order might be random (we can't easily tell without further investigation if GitHub changed the order of a PATH).
2. If py2exe knows it's building a 32-bit executable then it should take steps to not copy any 64-bit DLLs, and I assume vice versa.

Edit: One of these days I'll remember that GitHub issues need the closing triple-backtick on its own line.

**Comment by albertosottile** [comment]
Hello, thanks for reporting this. As you can see from [here](https://github.com/py2exe/py2exe/blob/93c4d9ea21c0b3da86b5008fb2f7785b7934af81/.appveyor.yml#L152), this is a known issue of the new `dllfinder`. **TL,DR**: If you want to fix this on your side,I would advise you to do the same and/or adjust `PATH` to fix this issue in your CI.

While preparing a reply, I had started to write a very long and detailed answer to your concerns and I actually had an idea on how to fix this bug. I guess that at the moment you do have a CI setup that reliably reproduces this bug, am I right? If so, would you be willing to install a test wheel with a fix for this issue in this setup, and confirm if the issue is solved or not?

If this is the case, I will prepare a "tentative" fix for this issue and provide you the link to the relative wheels in the next days.

**Comment by Athanasius** [comment]
Prior to you taking over py2exe on PyPi we'd been using a requirements.txt form:

```
git+https://github.com/albertosottile/py2exe.git@v0.10.0.2#egg=py2exe; sys_platform == 'win32'
```
so, if you can provide the tentative-fix wheel in that form it will be easy to edit our requirements-dev.txt to test it.  Else we'll have to put in a little more work in order to have the CI flow download a file and install from that.

**Comment by albertosottile** [comment]
I think https://dl.bintray.com/py2exe/py2exe/ should be usable for this purpose but, worst case scenario, I can make a pre-release on GitHub and then delete it afterwards...

**Comment by albertosottile** [comment]
@Athanasius Freshly built wheels for version 0.10.2.0 are available on https://bintray.com/py2exe/py2exe/py2exe/0.10.2.0#files and on https://dl.bintray.com/py2exe/py2exe/, and include a new `DllFinder` on steroids. 

My CI flow passes without manually deleting the rogue sqlite3 library, which gives me hope. On the negative side, it seems that now the finder is picking up **too much stuff**, e.g. a lot of `C:\Program Files\PowerShell\7\api-ms-win-*.dll` files.

Please let me know if these wheels also fix your issue. If so, it would remain to be seen if we want to exclude a priori all these `api-ms-win-*.dll` files from all possible bundles or not.

**Comment by Athanasius** [comment]
In our case it seems to be `C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-process-l1-1-0.dll` and friends, so the same stuff but on a different path.  They're mostly moot to us because the way we build our .msi for distribution just doesn't consider them so they never make it that far.

So, additional files aren't so harmful to us, but getting the wrong file definitely is.

I or @A-UNDERSCORE-D will run some tests with the tweaked wheels tomorrow.

**Comment by Athanasius** [comment]
1. The github workflow output shows it only copying libcrypt from the python directory, not again from Amazon.
2. No mention at all of that Amazon path in fact.
3. The resulting .msi file does work, yielding a working installed application.

In full our "COPY" output is now (NB: this is in a branch forked from our `develop` so using Python 3.8):

```
Copy c:\hostedtoolcache\windows\python\3.8.6\x86\python38.dll to dist.win32
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\select.pyd to dist.win32\select.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_bz2.pyd to dist.win32\_bz2.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_lzma.pyd to dist.win32\_lzma.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\pyexpat.pyd to dist.win32\pyexpat.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\unicodedata.pyd to dist.win32\unicodedata.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_socket.pyd to dist.win32\_socket.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_hashlib.pyd to dist.win32\_hashlib.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_ssl.pyd to dist.win32\_ssl.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_ctypes.pyd to dist.win32\_ctypes.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_queue.pyd to dist.win32\_queue.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_sqlite3.pyd to dist.win32\_sqlite3.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_multiprocessing.pyd to dist.win32\_multiprocessing.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_decimal.pyd to dist.win32\_decimal.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_asyncio.pyd to dist.win32\_asyncio.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_overlapped.pyd to dist.win32\_overlapped.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_elementtree.pyd to dist.win32\_elementtree.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_tkinter.pyd to dist.win32\_tkinter.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_testcapi.pyd to dist.win32\_testcapi.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\winsound.pyd to dist.win32\winsound.pyd
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-string-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-process-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-convert-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-synch-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-heap-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-locale-l1-1-0.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\libcrypto-1_1.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-string-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-com-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-localization-l1-2-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-sysinfo-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-winrt-error-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-environment-l1-1-0.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\tk86t.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-math-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-registry-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-security-base-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-heap-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-string-l2-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-version-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-file-l1-1-0.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\tcl86t.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-debug-l1-1-0.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\libssl-1_1.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-winrt-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-threadpool-l1-2-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-errorhandling-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\PowerShell\7\api-ms-win-core-delayload-l1-1-0.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\VCRUNTIME140.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-crt-utility-l1-1-0.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-processenvironment-l1-1-0.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\libffi-7.dll to dist.win32\
Copy DLL C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI\api-ms-win-core-processthreads-l1-1-0.dll to dist.win32\Windows
```

So, yes, lots of that api-ms-*.dll copying, but as I said, this doesn't actually impact us because we hard-code which files get built into the installer and these get ignored.

**Comment by albertosottile** [comment]
Thanks for running this, I see this result as an absolute win already! 😊 I have some ideas on how I can try to avoid picking unwanted libraries... Would you be willing to stick with me and test another wheel in the next days?

**Comment by Athanasius** [comment]
Sure, we can just re-use this same branch for further testing.

**Comment by albertosottile** [comment]
So, after digging this further, it seems to me that there is no clean way to determine if those libraries should be included or not just from the `DllFinder` results alone. That being said, looking at Microsoft's [documentation] (https://docs.microsoft.com/en-us/cpp/windows/universal-crt-deployment?view=msvc-160):

> Local deployment of the Universal CRT is supported, but not recommended for both performance and security reasons. 

Therefore, I would be inclined to filter out all the files that match the pattern suggested by them (`api-ms-win-*.dll`) from the bundle.

A special case is local deployment of the Visual C++ Redistributable library (`VCRUNTIME140.dll`) which is [also not recommended](https://docs.microsoft.com/en-us/cpp/windows/choosing-a-deployment-method?view=msvc-160) by Microsoft, but is actually done by a lot of developers, including Python's (`py2exe` typically takes this file from the root folder of the Python installation). Personally, as discussed in #14 and as you @Athanasius also do, I would also not include this library in my bundles.

In summary, I decided to filter out all these files (`api-ms-win-*.dll` and `VCRUNTIME*.dll`) from the `DllFinder`. These libraries will never be included in the bundles generated by `py2exe`. If this causes issues in running the packed executable, developers should redirect their users to Windows Update and/or the relevant standalone redistributable installers.

@Athanasius Wheels in the URL provided before now embed this change. Could you run your test again and verify that a) those libraries are **not** picked up anymore and b) your application still runs? Thanks

**Comment by Athanasius** [comment]
> In summary, I decided to filter out all these files (`api-ms-win-*.dll` and `VCRUNTIME*.dll`) from the `DllFinder`. These libraries will never be included in the bundles generated by `py2exe`. If this causes issues in running the packed executable, developers should redirect their users to Windows Update and/or the relevant standalone redistributable installers.

I agree with this conclusion.  I'd already stopped trying to ship the VCRUNTIME*.dll when checking MS docs pointed to that not being how it was meant to get onto user systems any more (Windows Update should take care of it).

I'll check the latest wheels tomorrow UK time.

**Comment by Athanasius** [comment]
Yup, that latest version results in a working package for us, without having copied in all those extraneous DLLs.

```
Copy c:\hostedtoolcache\windows\python\3.8.6\x86\python38.dll to dist.win32
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_bz2.pyd to dist.win32\_bz2.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_lzma.pyd to dist.win32\_lzma.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\select.pyd to dist.win32\select.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\pyexpat.pyd to dist.win32\pyexpat.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\unicodedata.pyd to dist.win32\unicodedata.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_socket.pyd to dist.win32\_socket.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_hashlib.pyd to dist.win32\_hashlib.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_ssl.pyd to dist.win32\_ssl.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_ctypes.pyd to dist.win32\_ctypes.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_queue.pyd to dist.win32\_queue.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_sqlite3.pyd to dist.win32\_sqlite3.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_multiprocessing.pyd to dist.win32\_multiprocessing.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_decimal.pyd to dist.win32\_decimal.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_asyncio.pyd to dist.win32\_asyncio.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_overlapped.pyd to dist.win32\_overlapped.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_elementtree.pyd to dist.win32\_elementtree.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_tkinter.pyd to dist.win32\_tkinter.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\_testcapi.pyd to dist.win32\_testcapi.pyd
Copy C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\winsound.pyd to dist.win32\winsound.pyd
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\libcrypto-1_1.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\tcl86t.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\tk86t.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\sqlite3.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\libffi-7.dll to dist.win32\
Copy DLL C:\hostedtoolcache\windows\Python\3.8.6\x86\DLLs\libssl-1_1.dll to dist.win32\
Copy lib file C:\hostedtoolcache\windows\Python\3.8.6\x86\lib\site-packages\certifi\cacert.pem to dist.win32\
```

**Comment by albertosottile** [comment]
Fixed in [0.10.2.0](https://github.com/py2exe/py2exe/releases/tag/v0.10.2.0).
